### ISSUE TITLE
2.9.10: UnboundLocalError: local variable 'module_style' referenced before assignment

### ISSUE THREAD

**Issue by resmo** [ansible/ansible#70168]
##### SUMMARY

after upgrade from 2.9.9 to 2.9.10 we see an exception in a task using ~~lineinfile~~ `delegate_facts=true`

##### ISSUE TYPE
- Bug Report

##### ANSIBLE VERSION

```paste below
2.9.10
```

##### CONFIGURATION

```paste below
ANSIBLE_NOCOWS(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = True
ANSIBLE_PIPELINING(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = True
ANSIBLE_SSH_ARGS(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = -o ControlMaster=auto -o ControlPersist=60s -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no
CACHE_PLUGIN(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = jsonfile
CACHE_PLUGIN_CONNECTION(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = .cache/facts
CACHE_PLUGIN_TIMEOUT(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = 86400
COLLECTIONS_PATHS(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = [u'/home/rmoser/Projects/puzzle/ansible-puzzle/collections_galaxy']
CONDITIONAL_BARE_VARS(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = False
DEFAULT_BECOME(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = True
DEFAULT_BECOME_ASK_PASS(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = False
DEFAULT_BECOME_METHOD(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = sudo
DEFAULT_BECOME_USER(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = root
DEFAULT_FORKS(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = 20
DEFAULT_GATHERING(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = smart
DEFAULT_HOST_LIST(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = [u'/home/rmoser/Projects/puzzle/ansible-puzzle/inventories/cloud_staging']
DEFAULT_LOAD_CALLBACK_PLUGINS(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = False
DEFAULT_MANAGED_STR(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = Warning: File is managed by Ansible
DEFAULT_PRIVATE_KEY_FILE(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = /home/rmoser/.ssh/id_rsa
DEFAULT_REMOTE_USER(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = ansible
DEFAULT_ROLES_PATH(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = [u'/home/rmoser/Projects/puzzle/ansible-puzzle/roles_galaxy', u'/home/rmoser/Projects/puzzle/ansible-puzzle/roles']
DEFAULT_STDOUT_CALLBACK(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = yaml
HOST_KEY_CHECKING(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = True
INTERPRETER_PYTHON(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = auto
RETRY_FILES_ENABLED(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = False

```

##### STEPS TO REPRODUCE

```yaml
---
- hosts: my.example.com
  gather_facts: yes
  tasks:
  - name: make something
    delegate_facts: True
    delegate_to: other.example.com
    lineinfile:
      path: /tmp/foo.txt
      regexp: '^Listen '
      insertafter: '^#Listen '
      line: Listen 8080

```

##### ACTUAL RESULTS

```paste below
The full traceback is:
Traceback (most recent call last):
  File "/home/rmoser/Projects/resmo/ansible/lib/ansible/executor/task_executor.py", line 146, in run
    res = self._execute()
  File "/home/rmoser/Projects/resmo/ansible/lib/ansible/executor/task_executor.py", line 654, in _execute
    result = self._handler.run(task_vars=variables)
  File "/home/rmoser/Projects/resmo/ansible/lib/ansible/plugins/action/normal.py", line 46, in run
    result = merge_hash(result, self._execute_module(task_vars=task_vars, wrap_async=wrap_async))
  File "/home/rmoser/Projects/resmo/ansible/lib/ansible/plugins/action/__init__.py", line 831, in _execute_module
    (module_style, shebang, module_data, module_path) = self._configure_module(module_name=module_name, module_args=module_args, task_vars=task_vars)
  File "/home/rmoser/Projects/resmo/ansible/lib/ansible/plugins/action/__init__.py", line 244, in _configure_module
    return (module_style, module_shebang, module_data, module_path)
UnboundLocalError: local variable 'module_style' referenced before assignment
fatal: [server.example.com]: FAILED! => 
  msg: Unexpected failure during module execution.
  stdout: ''

```

**Comment by mkrizek** [comment]
I couldn't reproduce this with an arbitrary lineinfile task. We'll need you to provide a minimal reproducer.

**Comment by resmo** [comment]
@mkrizek I added a reproducer.

**Comment by mkrizek** [comment]
Sounds like this might be related to #69604.

cc @bcoca

**Comment by srgvg** [comment]
Can confirm, we have the same problem, though it only triggers in particular case (localhost -> localhost)

Playbook:
```
# this might seem as a complex construct for just runnig setup/facts gathering on the `all` group,                        
# but this way, we are sure to have all the facts of all the hosts even if the play was run on a `--limit`ed set of hosts!
- name: Setup facts
  hosts: localhost
  tasks:
  ┊ - name: Gather facts from hosts
  ┊ ┊ setup:
  ┊ ┊ ┊ gather_subset: "all"
  ┊ ┊ delegate_to: "{{ item }}"
  ┊ ┊ delegate_facts: True
  ┊ ┊ with_items: "{{ groups['all'] }}"
  tags:
  ┊ - always

```

Results:

![20200619143809](https://user-images.githubusercontent.com/382239/85133538-2a41ee80-b23b-11ea-9fd9-f0b138aca4ee.png)

**Comment by mkrizek** [comment]
Since we now pass `use_vars` instead of `task_vars` to `modify_module`, we need to update information in `use_vars` for the next iteration too. The following makes the problem go away:
```diff
diff --git a/lib/ansible/plugins/action/__init__.py b/lib/ansible/plugins/action/__init__.py
index 83c7715d3e..62a185a3fd 100644
--- a/lib/ansible/plugins/action/__init__.py
+++ b/lib/ansible/plugins/action/__init__.py
@@ -249,7 +249,9 @@ class ActionBase(with_metaclass(ABCMeta, object)):
                     # store in local task_vars facts collection for the retry and any other usages in this worker
                     if use_vars.get('ansible_facts') is None:
                         task_vars['ansible_facts'] = {}
+                        use_vars['ansible_facts'] = {}
                     task_vars['ansible_facts'][discovered_key] = self._discovered_interpreter
+                    use_vars['ansible_facts'][discovered_key] = self._discovered_interpreter
                     # preserve this so _execute_module can propagate back to controller as a fact
                     self._discovered_interpreter_key = discovered_key
                 else:
```
The similar would need to be changed in the `else` branch as well. I am not sure if changing `task_vars` is needed though, or if changing all occurrences of `task_vars` to `use_vars` would be the fix.

**Comment by samdoran** [comment]
Reopening since the fix for this was reverted in #70272.
